### PR TITLE
 feature addition: [user_ldap] update user profile from LDAP

### PR DESCRIPTION
**Pull request by march42**
Updated documentation for [feature addition issue 32085](https://github.com/nextcloud/server/issues/32085)
[corresponding PR for server](https://github.com/nextcloud/server/pull/36565)
Added screenshot of additional section in LDAP/AD advanced tab.
Added explanations of LDAP/AD attribute settings and handling.

### CODE REVIEW

File: admin_manual/configuration_user/user_auth_ldap.rst
Comment (march42) [inline admin_manual/configuration_user/user_auth_ldap.rst:472]: This change is not mine!
The $home is still there, but those three lines in the documentation were removed somewhen between April 2022 and February 2023.

```suggestion
```

Should I kick this from my pull request? I'm really unsure about this.

File: admin_manual/configuration_user/user_auth_ldap.rst
Comment (come-nc) [inline admin_manual/configuration_user/user_auth_ldap.rst:524]: With or without the `@`?

File: admin_manual/configuration_user/user_auth_ldap.rst
Comment (march42) [reply admin_manual/configuration_user/user_auth_ldap.rst:488]: You're right and yet In my experience the values were always returned in the order of addition. So, I have choosen not to worry about it, because sorting of values or anything else would burn server ressource. Instead I have modified my schema and set attributes to SINGLE-VALUE.
Also. The eMail is handled likewise. First value gets primary eMail and additional values get additional eMail address. And it seems to be no problem.

File: admin_manual/configuration_user/user_auth_ldap.rst
Comment (march42) [reply admin_manual/configuration_user/user_auth_ldap.rst:485]: > So the user can change the fields, but the values get overridden?

Yes.

> How is it done for other fields sync from LDAP like display name and password, they do not appear as modifiable in the UI right? Would that be possible for profile field as well?

Those were never meant to be modified by the user.
I think it should be possible to change the user profile On/Off to On/Off/ReadOnly. Also, I don't think it is neccessary, to complicate this that way.

File: admin_manual/configuration_user/user_auth_ldap.rst
Comment (march42) [reply admin_manual/configuration_user/user_auth_ldap.rst:482]: This would need to publishing the protected functions sanitizePhone/sanitizeWebsite from AccountManager, because those are not accessible from user land code.

File: admin_manual/configuration_user/user_auth_ldap.rst
Comment (march42) [reply admin_manual/configuration_user/user_auth_ldap.rst:495]: No. If attribute name is not present in configuration or attribute is not present in the LDAP result, the user profile is ignored.
So, the users without admin defined values in LDAP can modify the profile and set what they like.
Otherwise the LDAP data should be written to user settings and user profile only updated/removed, after LDAP data changes. At least in my opinion.

File: admin_manual/configuration_user/user_auth_ldap.rst
Comment (march42) [reply admin_manual/configuration_user/user_auth_ldap.rst:524]: > With or without the `@`?

Actually Nextcloud doesn't bother or sanitize.
I think it should always be with, because for Twitter the `@` is defined as part of the handle `@username`
Same for Fediverse where the account name is defined to be `@name@instance`

File: admin_manual/configuration_user/user_auth_ldap.rst
Comment (march42) [reply admin_manual/configuration_user/user_auth_ldap.rst]: @come-nc I refined the code again, for better handling.
The profile data gets updated only, if the LDAP data has changed. I calculate and store a checksum of the profileValues from LDAP. If the checksum matches, the profile update gets skipped. Secondly, the user can now change his profile without it getting overwritten instantly. Thirdly, I write the checksum to cache and skip until the cached value expired.
Hope you like it.

File: admin_manual/configuration_user/user_auth_ldap.rst
Comment (march42) [reply admin_manual/configuration_user/user_auth_ldap.rst]: I removed it.

File: admin_manual/configuration_user/user_auth_ldap.rst
Comment (come-nc) [reply admin_manual/configuration_user/user_auth_ldap.rst:491]: > No. If attribute name is not present in configuration or attribute is not present in the LDAP result, the user profile is ignored. So, the users without admin defined values in LDAP can modify the profile and set what they like. Otherwise the LDAP data should be written to user settings and user profile only updated/removed, after LDAP data changes. At least in my opinion.

For me this is not the same thing.
When the attribute name is empty in configuration yes, it should be left alone.

But if the attribute name is set and its value is empty(unset) in LDAP, it should empty the value in Nextcloud.

Otherwise, this happens:

1. Admin maps phone number to field phoneNumber
2. User Bob gets a phone, admin sets phoneNumber in LDAP to the phone number
3. Nextcloud sync that to phone number field
4. Bob loses phone, admin removes phone number from LDAP
5. Nextcloud ignores this changes and let the outdated data in the profile :warning:

Comment (march42) [reply admin_manual/configuration_user/user_auth_ldap.rst:491]: Okay. I see your point @come-nc and I'll fix it this weekend.
The fix could be populating the array with empty values. I had this and removed it for reasons I can't remember.

Comment (march42) [reply admin_manual/configuration_user/user_auth_ldap.rst:491]: > I had this and removed it for reasons I can't remember.

1. User marc has his phone number manually set in profile
2. Admin configures in LDAP marc has no phone number

**Comment by march42** [comment]
> ❓ Questions

Do you think it's okay or did I miss something?  @come-nc

**Review by come-nc — APPROVED** [review]

**Comment by come-nc** [comment]
> > question Questions
> 
> Do you think it's okay or did I miss something? @come-nc

I think it’s good, could you try rebasing on master to see if it helps with CI?

**Comment by march42** [comment]
> I think it’s good, could you try rebasing on master to see if it helps with CI?

Done. And it now shows 'This workflow requires approval from a maintainer.'

**Comment by come-nc** [comment]
Thank you for your contribution and patience :blue_heart:

**Pull request merged**
